### PR TITLE
send record fixes to peers who send outdated records

### PR DESCRIPTION
**Pull request by whyrusleeping**
Peers who send records with expired validity are currently not sent a 'fixup' record after a peer does a get operation. This makes invalid records just return a blank record so that the later fixup logic can send them a better record.

License: MIT
Signed-off-by: Jeromy <jeromyj@gmail.com>

### CODE REVIEW

File: routing/dht/dht.go [hunk 1]
```diff
@@ -173,7 +173,9 @@ func (dht *IpfsDHT) getValueOrPeers(ctx context.Context, p peer.ID,
 		err = dht.verifyRecordOnline(ctx, record)
 		if err != nil {
 			log.Info("Received invalid record! (discarded)")
-			return nil, nil, err
+			// still return a non-nil record to signify that we received
+			// a bad record from this peer
+			record = new(pb.Record)
```
Comment (jbenet) [inline routing/dht/dht.go]: ooof-- that embeds a weird/non-obvious assumption here. what about returning a special error, and catching it later

File: routing/dht/routing.go [hunk 2]
```diff
@@ -170,6 +172,14 @@ func (dht *IpfsDHT) GetValues(ctx context.Context, key key.Key, nvals int) ([]ro
 
 		rec, peers, err := dht.getValueOrPeers(ctx, p, key)
 		if err != nil {
+			if err == routing.ErrNotFound {
+				// in this case, they responded with nothing,
+				// still send a notification
```
Comment (jbenet) [inline routing/dht/routing.go]: maybe add _why_ we want to send a notif here.

File: routing/dht/dht.go (hunk 1 above)
Comment (jbenet) [inline routing/dht/dht.go]: also, an invalid record should not necessarily be the same as an empty record, malicious peers who send invalid stuff should maybe not be talked to, etc. we may want to deal with them specially later (distinct from a valid but empty record)

Comment (whyrusleeping) [inline routing/dht/dht.go]: yeah, i agree here. Finding a clean way to do this is hard... :/

**Pull request merged**